src/CPULSSolver.cpp

```diff
@@ -355,177 +360,197 @@ void CPULSSolver::tallyLSScalarFlux(segment* curr_segment, int azim_index,
+    FP_PRECISION src_flat[_num_groups] 
+                 __attribute__ ((aligned(VEC_ALIGNMENT)));
+    FP_PRECISION src_linear[_num_groups] 
+                 __attribute__ ((aligned(VEC_ALIGNMENT)));
 
-      // Compute indexes upfront for performance
-      long scalar_idx = start_scalar_idx + e;
-      long first_linear_idx = start_linear_idx + 3*e;
-      int flat_idx = 2*e;
-
-      // Compute the flat and linear components of the source
-      sources[flat_idx] = 0.0;
-      sources[flat_idx+1] = 0.0;
-      for (int i=0; i<3; i++) {
-        int linear_idx = first_linear_idx + i;
-        sources[flat_idx] += _reduced_sources_xyz[linear_idx] * center_x2[i];
-        sources[flat_idx+1] += _reduced_sources_xyz[linear_idx] * direction[i];
-      }
-      sources[flat_idx] += _reduced_sources[scalar_idx];
+#pragma omp simd aligned(src_flat, src_linear)
+    for (int e=0; e < _num_groups; e++) {
+      src_flat[e] = _reduced_sources(fsr_id, e);
+      for (int i=0; i<3; i++)
+        src_flat[e] += _reduced_sources_xyz(fsr_id, e, i) * center_x2[i];
+      src_linear[e] = _reduced_sources_xyz(fsr_id, e, 0) * direction[0];
+      src_linear[e] += _reduced_sources_xyz(fsr_id, e, 1) * direction[1];
+      src_linear[e] += _reduced_sources_xyz(fsr_id, e, 2) * direction[2];
     }
 
+    // Compute the flux attenuation and tally contribution
+#pragma omp simd aligned(tau, src_flat, src_linear, exp_F1, exp_F2, exp_H, fsr_flux, fsr_flux_x, fsr_flux_y, fsr_flux_z)
     for (int e=0; e < _num_groups; e++) {
 
-      // Load sources
-      int first_idx = 2*e;
-      FP_PRECISION src_flat = sources[first_idx];
-      FP_PRECISION src_linear = sources[first_idx+1];
-
-      // Load exponential terms
-      first_idx += e; // equivalent to 3*e
-      FP_PRECISION exp_F1 = exponentials[first_idx];
-      FP_PRECISION exp_F2 = exponentials[first_idx+1];
-      FP_PRECISION exp_H = exponentials[first_idx+2];
-      FP_PRECISION tau = sigma_t[e] * length_2D;
-
       // Compute the change in flux across the segment
-      exp_H *= length * track_flux[e] * tau * wgt;
-      FP_PRECISION delta_psi = (tau * track_flux[e] - length_2D * src_flat) *
-          exp_F1 - src_linear * length_2D * length_2D * exp_F2;
+      exp_H[e] *= length * track_flux[e] * tau[e] * wgt;
+      FP_PRECISION delta_psi = (tau[e] * track_flux[e] - length_2D * src_flat[e])
+           * exp_F1[e] - src_linear[e] * length_2D * length_2D *
+            exp_F2[e];
+      track_flux[e] -= delta_psi;
+      delta_psi *= wgt;
 
       // Increment the fsr scalar flux and scalar flux moments
-      first_idx += e; // equivalent to 4*e
-      fsr_flux[first_idx] += wgt * delta_psi;
-      first_idx++;
-      for (int i=0; i<3; i++)
-        fsr_flux[first_idx + i] += exp_H * direction[i] + wgt * delta_psi *
-                                    position[i];
-
-      // Decrement the track flux
-      track_flux[e] -= delta_psi;
+      fsr_flux[e] += delta_psi;
```

Comment:
So if -NGROUPS=70 is set, then we don't need it for SIMD, the compiler picks it up.
At least for gnu 6.2

The tau loops also have this issue, because of the % _num_groups